doc/designs/quic-design/connection-id-cache.md

```diff
@@ -4,252 +4,164 @@ QUIC Connection ID Cache
 Connection ID Cache
 -------------------
 
-A table matching Connection IDs with Connection objects represented
-via SSL objects.
+The connection ID cache is responsible for managing connection IDs, both local
+and remote.
 
-_In MVP there is a many-to-1 matching of Connection IDs to Connection
-objects.  Refer third paragraph in [5.1]_
+### Remote Connection IDs
+
+For remote connection IDs, we need to be able to:
+
+* add new IDs per connection;
+* pick a non-retired ID associated from those available for a connection and
+* select a connection ID by sequence number and retire that and all older IDs.
+
+The cache will be implemented as a double ended queue as part of the
+QUIC_CONNECTION object.  There is no requirement to maintain a global
+mapping since remote IDs are only used when sending packets, not
+receiving them.
+
+In MVP, a many-to-1 matching of Connection IDs per Connection object
+is required.  Refer third paragraph in [5.1].
+
+When picking a non-retired connection ID, the youngest available will
+be chosen.
+
+### Local Connection IDs
+
+For local connection IDs, we need to be able to:
+
+* generate a new connection ID and associate it with a connection;
+* query if a connection ID is present;
+* for a server, map a connection ID to a QUIC_CONNECTION;
+* drop all connection IDs associdated with a QUIC_CONNECTION and
+* select a connection ID by sequence number and retire that and all older IDs.
+
+All connection IDs issued by our stack must be the same length because
+short form packets include the connection ID but no length byte.  Random
+connection IDs of this length will be allocated.  Note that no additional
+information will be contained in the connection ID.
+
+There will be one global set of local connection IDs which is shared
+by all connections over all SSL_CTX objects.  This is used to dispatch
+incoming datagrams to their correct destination and will be implemented
+as a dictionary.
+
+#### Notes
+
+* For MVP, it would be sufficient to only use a zero length connection ID.
+* For MVP, a connection ID to QUIC_CONNECTION mapping need not be implemented.
+* Funnelling all received packets through a single socket is likely to be
+  bottleneck once we support multiple connections.
+  An alternative would be receiving from <host address, source port> pairs.
+* For MVP, the local connection ID cache need only have one element.
+  I.e. there is no requirement to implement any form of lookup.
 
 API
 ---
 
+QUIC connection IDs are defined in #18949 but and extra fields and some
+extra functions are available:
+
 ```c
-#define QUIC_MAX_ID_LENGTH  20
+/* QUIC connection ID representation. */
+#define QUIC_MAX_CONN_ID_LEN   20
+
+typedef struct quic_conn_id_st {
+    unsigned char id_len;
+    unsigned char id[QUIC_MAX_CONN_ID_LEN];
+    uint64_t      seq_no;
+    unsigned int  retired : 1;
+} QUIC_CONN_ID;
```

Comment:
Please do not add fields to this structure. It is already used in various places including packet serialization code which have no need for these extra fields. Please define a new structure which wraps the existing one. `QUIC_CONN_ID_DESC`, maybe.

Reply:
I figured the same thing out last night...

Reply:
Although I'm dubious about the sequence number needing to be separate.  It is isomorphic to the connection ID and IMO part of it.